docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,260 @@
+(`string.join`), an underscore could be used instead (`string_join`).
+
+Pros:
+
+* Clearer when something is in the standard library compared to being a
+  reference to a component.
+* Not possible for a standard library namespace to collide with a component
+  namespace.
+
+Cons:
+
+* Inconsistent with `constants` which uses dot-separation for retrieving a
+  specific constant.
+    * This could be mitigated by changing `constants.os` to `constants_os` or
+      similar.
+
+### Do not standardize existing identifiers
+
+Alternatively, a pattern for standardizing new identifiers could be
+established, while existing identifiers are left untouched.
+
+Pros:
+
+* The majority of the existing standard library remains valid.
+
+Cons:
+
+* Until the number of new identifiers is far more than the number of current
+  identifiers, the standard library will appear very inconsistent, with a heavy
+  mix of namespaced and non-namespaced identifiers.
+* An existing identifier may belong in a namespace alongside new identifiers,
+  but is left out for backwards compatibility, causing inconsistencies.
+
+### Do not establish a standardization for identifiers
+
+Alternatively, we can omit standardizing identifiers at all, and continue to
+let the standard library grow organically.
+
+Pros:
+
+* Less work for contributors :)
+
+Cons:
+
+* It will be harder to understand and navigate the standard library as the
+  number of identifiers grows.
+
+## Compatibility
+
+This proposal is properly backwards compatible, with the old identifiers being
+marked deprecated in favour of their namespaced equivalents (where a namespace
+was introduced).
+
+It is possible that any namespaces introduced in this proposal may collide with
+an existing custom component. Custom components shadow the stdlib, so this is
+not a breaking change, but users will have to change the name of their custom
+components that collide with a stdlib namespace before being able to use
+functions in that namespace.
+
+By default, the removal deprecated identifiers would be considered for an
+eventual 2.0 release, but they may be kept around longer based on usage.
+
+## Implementation
+
+The implementation of this proposal will be broken down into the following
+steps:
+
+1. Implement the new namespaced identifiers, and deprecate the old ones in
+   documentation.
+
+2. Update examples, tests, internal configs, and config converters to use the
+   new namespaced identifiers.
+
+2. Find a way to detect and report usage of deprecated identifiers to a user.
+
+3. Add standard library usage to usage stats, so usage of deprecated
+   identifiers can be tracked over time, allowing maintainers to make an
+   informed decision for if they can be removed alongside a major release.
+
+Additionally, as this change may introduce new instances where custom component
+names shadow a namespace, we should add a warning to notify the user of
+shadowing (if detected).
```

Comment:
Is shadowing an actual concern, given that stdlib functions are called using `()`, whereas component exports are not?

Reply:
I wouldn't say "concern" (in this case, shadowing is preferable), but it is real; the syntax evaluator treats all identifier lookups equally, regardless of whether they're called as a function. 

Plus, there are non-function identifiers in the stdlib, like `constants.hostname`.